src/paint.rs

```diff
@@ -0,0 +1,235 @@
+    }
+}
+
+impl Brush for crate::piet::Brush {
+    fn make_brush<'a>(
+        &'a self,
+        _piet: &mut Piet,
+        _shape: &impl Shape,
+    ) -> Cow<'a, crate::piet::Brush> {
+        Cow::Borrowed(self)
+    }
+}
+
+impl<'a, 'b: 'a> Deref for PaintCtx<'a, 'b> {
+    type Target = Piet<'b>;
+
+    fn deref(&self) -> &Self::Target {
+        self.render_ctx
+    }
+}
+
+impl<'a, 'b: 'a> DerefMut for PaintCtx<'a, 'b> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.render_ctx
+    }
+}
+
+// The block below expands the range of brush arguments to allow anything
+// implementing the brush trait. This is a signifcant ergonomics improvement.
+impl<'a, 'b> PaintCtx<'a, 'b> {
+    /// Fill a shape.
+    pub fn fill(&mut self, shape: impl Shape, brush: &impl Brush, fill_rule: FillRule) {
+        let concrete_brush = brush.make_brush(self, &shape);
+        self.render_ctx
+            .fill(shape, concrete_brush.borrow(), fill_rule);
+    }
+
+    /// Stroke a shape.
+    pub fn stroke(
+        &mut self,
+        shape: impl Shape,
+        brush: &impl Brush,
+        width: f64,
+        style: Option<&StrokeStyle>,
+    ) {
+        let concrete_brush = brush.make_brush(self, &shape);
+        self.render_ctx
+            .stroke(shape, concrete_brush.borrow(), width, style);
+    }
+
+    // Note: we should make `draw_text` participate too, but it requires the bounding box
+    // for gradient creation, and we don't have a good way to do that yet.
+}
+
+impl GradientStops for Vec<GradientStop> {
+    fn to_vec(self) -> Vec<GradientStop> {
+        self
+    }
+}
+
+impl<'a> GradientStops for &'a [GradientStop] {
+    fn to_vec(self) -> Vec<GradientStop> {
+        self.to_owned()
+    }
+}
+
+// Generate equally-spaced stops.
+impl<'a> GradientStops for &'a [Color] {
+    fn to_vec(self) -> Vec<GradientStop> {
+        if self.is_empty() {
+            Vec::new()
+        } else {
+            let denom = (self.len() - 1).max(1) as f32;
+            self.iter()
+                .enumerate()
+                .map(|(i, c)| GradientStop {
+                    pos: (i as f32) / denom,
+                    color: c.to_owned(),
+                })
+                .collect()
+        }
+    }
+}
+
+impl<'a> GradientStops for (Color, Color) {
+    fn to_vec(self) -> Vec<GradientStop> {
+        let stops: &[Color] = &[self.0, self.1];
+        GradientStops::to_vec(stops)
+    }
+}
+
+impl<'a> GradientStops for (Color, Color, Color) {
+    fn to_vec(self) -> Vec<GradientStop> {
+        let stops: &[Color] = &[self.0, self.1, self.2];
+        GradientStops::to_vec(stops)
+    }
+}
+
+impl<'a> GradientStops for (Color, Color, Color, Color) {
+    fn to_vec(self) -> Vec<GradientStop> {
+        let stops: &[Color] = &[self.0, self.1, self.2, self.3];
+        GradientStops::to_vec(stops)
+    }
+}
+
+impl Gradient {
+    pub fn new(start: UnitPoint, end: UnitPoint, stops: impl GradientStops) -> Gradient {
```

Comment:
this is unclear enough to deserve a doc comment, ideally with a couple `# Examples`.